core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/ElementDestinationSelector.java

```diff
@@ -0,0 +1,143 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.index.IndexVariableDemand;
+import org.optaplanner.core.impl.domain.variable.index.IndexVariableSupply;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonListInverseVariableDemand;
+import org.optaplanner.core.impl.domain.variable.supply.SupplyManager;
+import org.optaplanner.core.impl.heuristic.selector.AbstractSelector;
+import org.optaplanner.core.impl.heuristic.selector.IterableSelector;
+import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelector;
+import org.optaplanner.core.impl.heuristic.selector.value.EntityIndependentValueSelector;
+import org.optaplanner.core.impl.solver.random.RandomUtils;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+
+/**
+ * Selects destinations for list variable change moves. The destination specifies a future position in a list variable,
+ * expressed as an {@link ElementRef}, where a moved element or subList can be inserted.
+ * <p>
+ * Destination completeness is achieved by using both entity and value child selectors.
+ * When an entity <em>A</em> is selected, the destination becomes <em>A[0]</em>.
+ * When a value <em>x</em> is selected, its current position <em>A[i]</em> is determined using inverse and index supplies and
+ * the destination becomes <em>A[i + 1]</em>.
+ * <p>
+ * Fairness in random selection is achieved by first deciding between entity and value selector with a probability that is
+ * proportional to the entity/value ratio. The child entity and value selectors are assumed to be fair.
```

Comment:
TODO prove this.

Reply:
What we know:
$P(E) = N_E/N$ - probability of choosing entity selector
$P(V) = N_V/N$ - probability of choosing value selector
$P(e) = 1/N_E$ - probability of entity selector returning the entity $e$
$P(v) = 1/N_V$ - probability of value selector returning the value $v$

What we're looking for:
$P(D_e)$ - unconditional probability of selecting destination `e[0]`
$P(D_v)$ - unconditional probability of selecting destination `x[i + 1]` where `x` is the inverse entity of `v` and `i` is index of `v`

I think that $P(D_e)$ is simply $P(E) \cdot P(e) = N_E/N \cdot 1/N_E = 1/N$ because selecting the entity $e$ cannot happen without choosing the entity selector first, so the conditional probability formula is not needed.

But even if I approach it with conditional probability I get the same result:
$P(D_e|E)=P(e)=1/N_E$ - probability of selecting destination `e[0]` under the condition that the entity selector was chosen. Equals the probability of entity selector returning entity $e$.
$P(E|D_e)=P(E|e)=1$ - probability that entity selector was chosen under the condition that destination `e[0]` was selected. It's 1 since `e[0]` cannot be selected without choosing entity selector.

Using Bayes theorem:

$P(D_e|E)=\frac{P(E|D_e) \cdot P(D_e)}{P(E)}$,

we're looking for the unconditional probability of selecting destination `e[0]` ( $P(D_e)$ ):

$P(D_e)=\frac{P(D_e|E) \cdot P(E)}{P(E|D_e)} = \frac{ \frac{1}{N_E} \cdot \frac{N_E}{N} }{1} = \frac{1}{N}$

Similarly for value-based destinations.